crates/tx/src/types.rs

```diff
@@ -1101,26 +1171,30 @@ impl Tx {
         None
     }
 
-    /// Set the transaction memo hash stored in the header
+    /// Set the last transaction memo hash stored in the header
     pub fn set_memo_sechash(&mut self, hash: namada_core::hash::Hash) {
-        self.header.memo_hash = hash;
-    }
+        let item = match self.header.batch.pop() {
+            Some(mut last) => {
+                last.memo_hash = hash;
+                last
+            }
+            None => TxCommitments {
+                memo_hash: hash,
+                ..Default::default()
+            },
+        };
 
-    /// Get the hash of this transaction's memo from the heeader
-    pub fn memo_sechash(&self) -> &namada_core::hash::Hash {
-        &self.header.memo_hash
+        self.header.batch.insert(item);
     }
 
-    /// Get the memo designated by the memo hash in the header
-    pub fn memo(&self) -> Option<Vec<u8>> {
-        if self.memo_sechash() == &namada_core::hash::Hash::default() {
+    /// Get the memo designated by the memo hash in the header for the specified
+    /// commitment
+    pub fn memo(&self, cmt: &TxCommitments) -> Option<Vec<u8>> {
```

Comment:
I've modified these methods to require a reference to the commitments that identify an inner tx. we could to this also with just an integer but I believe this is a bit safer and more clear in terms of how an inner tx is identified inside the batch. Sill we can change this logic if that's the case

Reply:
I like this!

Reply:
Opened #3200 to address this separately